zkevm-circuits/src/evm_circuit.rs

```diff
@@ -1 +1,598 @@
+    GasAvailable,
+    // Value in wei of call.
+    Value,
+    // If call succeeds or not in the future.
+    IsPersistant,
+    // If call is within a static call.
+    IsStatic,
+    // If call is `CREATE*`. We lookup op from calldata when is create,
+    // otherwise we lookup code under address.
+    IsCreate,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum CallStateField {
+    // Program counter.
+    ProgramCounter,
+    // Stack pointer.
+    StackPointer,
+    // Memory size.
+    MemeorySize,
+    // Gas counter.
+    GasCounter,
+    // State trie write counter.
+    StateWriteCounter,
+    // Last callee's unique identifier.
+    CalleeID,
+    // Offset of returndata.
+    ReturndataOffset,
+    // Size of returndata.
+    ReturndataSize,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum BusMappingLookup<F> {
+    // Read-Write
+    OpExecutionState {
+        field: CallStateField,
+        value: Expression<F>,
+        is_write: bool,
+    },
+    Stack {
+        // One can only access its own stack, so id is no needed to be
+        // specified.
+        // Stack index is always deterministic respect to stack pointer, so an
+        // index offset is enough.
+        index_offset: i32,
+        value: Expression<F>,
+        is_write: bool,
+    },
+    Memory {
+        // Some ops like CALLDATALOAD can peek caller's memeory as calldata, so
+        // we allow it to specify the id.
+        id: Expression<F>,
+        index: Expression<F>,
+        value: Expression<F>,
+        is_write: bool,
+    },
+    AccountStorage {
+        address: Expression<F>,
+        location: Expression<F>,
+        value: Expression<F>,
+        is_write: bool,
+    },
+    // TODO: AccountNonce,
+    // TODO: AccountBalance,
+    // TODO: AccountCodeHash,
+
+    // Read-Only
+    Call {
+        field: CallField,
+        value: Expression<F>,
+    },
+    Bytecode {
+        hash: Expression<F>,
+        index: Expression<F>,
+        value: Expression<F>,
+    },
+    // TODO: Block,
+    // TODO: Bytecode,
+    // TODO: Tx,
+    // TODO: TxCalldata,
+}
+
+#[derive(Clone, Copy, Debug, PartialOrd, PartialEq, Eq, Ord)]
+pub(crate) enum FixedLookup {
+    // Noop provides [0, 0, 0, 0] row
+    Noop,
+    // meaningful tags start with 1
+    Range256,
+    BitwiseAnd,
+    BitwiseOr,
+    BitwiseXor,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum Lookup<F> {
+    FixedLookup(FixedLookup, [Expression<F>; 3]),
+    BusMappingLookup(BusMappingLookup<F>),
+}
+
+#[derive(Clone, Debug)]
+struct Constraint<F> {
+    name: &'static str,
+    selector: Expression<F>,
+    linear_combinations: Vec<Expression<F>>,
+    lookups: Vec<Lookup<F>>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Cell<F> {
+    // expression for constraint
+    expression: Expression<F>,
+    // relative position to selector for synthesis
+    column: Column<Advice>,
+    offset: usize,
+}
+
+impl<F: FieldExt> Cell<F> {
+    fn exp(&self) -> Expression<F> {
```

Comment:
```suggestion
    fn expr(&self) -> Expression<F> {
```
Avoid the confusion with `exp` for exponential

Reply:
Thanks, `expr` sounds much better! Fixed in recent commit.